### PR TITLE
Wip/hughsie/fu install task

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

File: src/fu-main.c [hunk 1]
```diff
@@ -403,21 +412,233 @@ fu_main_authorize_install_cb (GObject *source, GAsyncResult *res, gpointer user_
...
+fu_main_error_array_build_composite (GPtrArray *errors)
+{
+	FwupdError err_prio[] =		{ FWUPD_ERROR_INVALID_FILE,
+					  FWUPD_ERROR_VERSION_SAME,
+					  FWUPD_ERROR_VERSION_NEWER,
+					  FWUPD_ERROR_NOT_SUPPORTED,
+					  FWUPD_ERROR_INTERNAL,
+					  FWUPD_ERROR_NOT_FOUND,
+					  FWUPD_ERROR_LAST };
+	FwupdError err_all_uptodate[] =	{ FWUPD_ERROR_VERSION_SAME,
```
Comment (superm1) [inline src/fu-main.c]: I'd think you probably also want to match VERSION_NEWER in this case right?

Comment (hughsie) [reply src/fu-main.c]: I don't think so. "All updatable firmware is already installed" isn't right if you're trying to install `firmware-1.2.3.cab` when you have 1.2.4 installed. Maybe we could add a new `err_all_newer` with  message like "All updatable firmware is already a newer version" or something like that?

Comment (superm1) [reply src/fu-main.c]: Yeah that's a good alternative.

File: src/fu-engine.c [hunk 2]
```diff
@@ -2060,6 +2049,14 @@ fu_engine_get_store_from_blob (FuEngine *self, GBytes *blob_cab, GError **error)
 	if (store == NULL)
 		return NULL;
 
+	/* fix all the apps */
+	apps = as_store_get_apps (store);
+	for (guint i = 0; i < apps->len; i++) {
+		AsApp *app = g_ptr_array_index (apps, i);
+		fu_engine_vendor_quirk_release_version (self, app);
```
[the remark this answers is not in the document]
Comment (hughsie) [reply src/fu-engine.c]: Sure can, fixed.

File: src/fu-main.c (hunk 1 above)
Comment (hughsie) [reply src/fu-main.c]: Fixed.

**Comment by superm1** [comment]
This gets it all almost right on my composite device.  I built the two cabs for my testing.  I think you can generate the same result even without a composite device.  Just group together metainfo for say a unifying firmware, UEFI firmware and TPM firmware.

First cab had:
* One device not marked updatable
* One device fully up to date
* one device needs to be updated

I ran install on that and it did what I expected, no errors shown, installed the update on the device that needed it, done.

Second cab had:
* One device not marked updatable
* Two devices fully up to date

I expected to see a message that all updatable devices have been updated but instead I saw:

> "Specified firmware is already installed 'd.c.b.a'"

For my third test I removed the device that's not updatable from the topology and used the second CAB.
I saw the expected message 

> "All updatable firmware is already installed".

**Comment by hughsie** [comment]
> I expected to see a message that all updatable devices have been updated

`Specified firmware is already installed` is what I thought would happen; if the device isn't updatable then it gets its error disregarded. The logic here is we don't want to include devices that match the GUID but the user can't actually interact with.

Maybe I'm missing some information; why would you include a firmware update in the .cab file if the device is not upgradable?

**Comment by superm1** [comment]
> Specified firmware is already installed is what I thought would happen; if the device isn't updatable then it gets its error disregarded. The logic here is we don't want to include devices that match the GUID but the user can't actually interact with.

I guess that makes sense.  I think what i'm expecting is that the error is actually disregarded in this situation rather than the best possible error picked (which is one device has firmware already installed).  I  expect it to behave more like my 3rd test above.

> Maybe I'm missing some information; why would you include a firmware update in the .cab file if the device is not upgradable?

For if fwupd doesn't yet support it but eventually will?  Think like someone running an older version of fwupd that didn't support that particular composite device.

**Comment by hughsie** [comment]
Fair point. Does that last commit fix things?

**Comment by superm1** [comment]
Yep, it's working exactly like I expect in all the scenarios I've tried now.

**Comment by hughsie** [comment]
Thanks for all the feedback and testing (and the initial patch) -- if we need to add ordering of `FuInstallTask` operations we can add that pretty easily when required. If you need this ASAP, yell.

**Comment by superm1** [comment]
Yeah if you can add that it will be quite useful for me. We're getting closer to needing to define order of updates for the stuff I'm doing with composite devices. My preference would be to let it be done via a plugin configurable policy of update parents before children and then set the device parents accordingly. I think that should work pretty well for unifying too right?

**Comment by superm1** [comment]
s,plugin configurable policy,device configurable policy,

As in set the policy when creating the device if parents go before it or children go before it.

**Pull request merged**